pennylane/ops/op_math/exp.py

```diff
@@ -144,14 +146,35 @@ def __repr__(self):
             else f"Exp({self.coeff} {self.base.name})"
         )
 
+    # pylint: disable=attribute-defined-outside-init
+    def __copy__(self):
+        # this method needs to be overwritten because the base must be copied too.
+        copied_op = object.__new__(type(self))
+        # copied_op must maintain inheritance structure of self
+        # Relevant for symbolic ops that mix in operation-specific components.
+
+        for attr, value in vars(self).items():
+            if attr not in {"_hyperparameters"}:
+                setattr(copied_op, attr, value)
+
+        copied_op._hyperparameters = copy(self.hyperparameters)
+        copied_op.hyperparameters["base"] = copy(self.base)
+        copied_op._data = copy(self._data)
+
+        return copied_op
+
+    @property
+    def hash(self):
+        return hash((str(self.name), self.base.hash, str(self.coeff)))
+
     @property
     def data(self):
-        return [[self.coeff], self.base.data]
+        return self._data
 
-    @data.setter
-    def data(self, new_data):
-        self.coeff = new_data[0][0]
-        self.base.data = new_data[1]
```

Comment:
The setter was causing problems.  

When the data is changed, PL tends to use `op.data[i] = new_value`, which bypasses the setter. It just mutates the returned list. I'm amazed this wasn't a problem earlier. But basically, we need to reconsider how we deal with operator parameters and data, but this is a good enough quick fix.

Reply:
I've added a data setter on `Exp` and on `Evolution`. I think the `Exp` one is pretty non-controversial, but I'd like some feedback on the one in `Evolution`. Right now it lets you set either just the primary parameter, or set both the primary parameter and the data for the base. This means it can take an argument that is of a different shape than the operator data:

```
h = qml.PauliX(0) + qml.PauliY(1)
op = Evolution(h, 2)

op.data
>>> [2]
op.base.data
>>> [1, 1]
```

```
op.data = [3]
op.data
>>> [3]
```

```
op.data = [[4], [2, 2]]
op.data
>>> [4]
op.base.data
>>> [2, 2]
```

Is it weird? Maybe you shouldn't be able to set the base data directly through Evolution.data?

Reply:
Ah, okay. I just saw Christina's comment. Should I remove them again?

I did notice this morning that setting data on Hamiltonian doesn't actually change the coefficients, only the data list. So if you simplify it after changing the data, for example, it uses the initial data, not the new data.